### PR TITLE
Decrease scope of lock in ServiceContainer; refactor ServiceContainer

### PR DESCRIPTION
**Pull request by SergeiPavlov**
`Dictionary<>` is thread-safe for read-only operations

### CODE REVIEW

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs [hunk 1]
```diff
@@ -74,38 +80,31 @@ protected override IEnumerable<object> HandleGetAll(Type serviceType)
...
-      try {
-        for (int i = 0; i < pInfos.Length; i++)
-          args[i] = Get(pInfos[i].ParameterType);
       }
-      finally {
-        creating.Remove(serviceInfo.Type);
+      var pInfos = cachedInfo.Second;
+
+      // Not very optimal, but...
+      lock (_lock) {
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/IoC/ServiceContainer.cs:91]: This lock is for access to `private readonly HashSet<Type> creating`  to prevent recursive dependencies
I don't see how we can avoid it without lock by simple way.
We can use Thread-local variables

[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/IoC/ServiceContainer.cs]: the lock is removed

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs [hunk 2]
```diff
@@ -86,25 +86,21 @@ protected virtual object CreateInstance(ServiceRegistration serviceInfo)
...
-        return cInfo.Invoke(args);
+      if (pInfos.Length == 0)
+        return Activator.CreateInstance(serviceInfo.MappedType);
+      var args = new object[pInfos.Length];
+      var key = (serviceInfo.Type, Thread.CurrentThread.ManagedThreadId);
+      if (!creating.TryAdd(key, true)) {
+        throw new ActivationException(Strings.ExRecursiveConstructorParemeterDependencyIsDetected);
       }
+      try {
+        for (int i = 0; i < pInfos.Length; i++)
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/IoC/ServiceContainer.cs:97]: Please add braces as you touched the code.

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs [hunk 3]
```diff
@@ -86,25 +86,21 @@ protected virtual object CreateInstance(ServiceRegistration serviceInfo)
...
-        creating.Add(serviceInfo.Type);
-        try {
-          for (int i = 0; i < pInfos.Length; i++)
-            args[i] = Get(pInfos[i].ParameterType);
-        }
-        finally {
-          creating.Remove(serviceInfo.Type);
-        }
-        return cInfo.Invoke(args);
+      if (pInfos.Length == 0)
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/IoC/ServiceContainer.cs:89]: Braces :)

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs [hunk 4]
```diff
@@ -311,9 +310,8 @@ public ServiceContainer(IEnumerable<ServiceRegistration> configuration = null, I
     public override void Dispose()
     {
       using (var toDispose = new DisposableSet()) {
-        foreach (var pair in instances) {
-          var service = pair.Value;
-          if (service is IDisposable disposable) {
+        foreach (var (_, lazy) in instances) {
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/IoC/ServiceContainer.cs:313]: Can we instead enumerate `instances.Values`?

**Review by AlexUstinov — APPROVED** [review]

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs (hunk 2 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/IoC/ServiceContainer.cs]: done

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs (hunk 4 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/IoC/ServiceContainer.cs]: done

**Review by botinko — APPROVED** [review]
I'd add some tests on edge cases like 
* Recursive Constructor Parameter Dependency
* Ensure, that singletons constructors called only once in case of parallel resolve
